File: C#/0173. Binary Search Tree Iterator.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class BSTIterator {
    
    Queue<int> rootQueue=new Queue<int>();
    public BSTIterator(TreeNode root) {
        Stack<TreeNode> rootStack=new Stack<TreeNode>();
        while(rootStack.Count()!=0 ||root!=null){
            while(root!=null){
                rootStack.Push(root);
                root=root.left;
            }
            root=rootStack.Pop();
            rootQueue.Enqueue(root.val);
            root=root.right;
        }
    }
        
    /** @return the next smallest number */
    public int Next() {
        return rootQueue.Dequeue();
    }
    
    /** @return whether we have a next smallest number */
    public bool HasNext() {
        return rootQueue.Count()>0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */

File: 20190519python/0173. Binary Search Tree Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.stack=[]
        self.readTree(root)
        
    def readTree(self,root):
        if not root:
            return
        if root.left:
            self.readTree(root.left)
        self.stack.append(root.val)
        if root.right:
            self.readTree(root.right)
            
    def next(self) -> int:
        """
        @return the next smallest number
        """
        return self.stack.pop(0)
        

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return len(self.stack)>0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
